python/packages/ai/teams/ai/planners/assistants_planner.py

```diff
@@ -0,0 +1,379 @@
+            if run.status == "requires_action":
+                if handle_actions:
+                    return run
+                break
+            if run.status == "cancelled":
+                return run
+            if run.status == "failed":
+                return run
+            if run.status == "completed":
+                return run
+            if run.status == "expired":
+                return run
+        return None
+
+    def _generate_plan_from_tools(self, state: TurnState, required_action: RequiredAction) -> Plan:
+        plan = Plan()
+        tool_map: Dict = {}
+        for tool_call in required_action.submit_tool_outputs.tool_calls:
+            tool_map[tool_call.function.name] = tool_call.id
+            plan.commands.append(
+                PredictedDoCommand(
+                    action=tool_call.function.name,
+                    parameters=json.loads(tool_call.function.arguments),
+                )
+            )
+        state.set(SUBMIT_TOOL_OUTPUTS_MAP, tool_map)
+        return plan
+
+    async def _generate_plan_from_messages(self, thread_id: str, last_message_id: str) -> Plan:
+        # Find the new messages
+        messages = await self._client.beta.threads.messages.list(thread_id)
+        new_messages = []
+        for message in messages.data:
+            if message.id == last_message_id:
+                break
+            new_messages.append(message)
+
+        # listMessages return messages in desc, reverse to be in asc order
+        new_messages.reverse()
+
+        # Convert the messages to SAY commands
+        plan = Plan()
+        for message in new_messages:
+            for content in message.content:
+                if content.type == "text":
+                    plan.commands.append(PredictedSayCommand(response=content.text.value))
+
+        return plan
+
+    async def _block_on_in_progress_runs(self, thread_id: str) -> None:
+        # We loop until we're told the last run is completed
+        while True:
+            run = await self._retrieve_last_run(thread_id)
+            if not run:
+                return
+            if self._is_run_completed(run):
+                return
+
+            # Wait for the current run to complete and then
+            # loop to see if there's already a new run.
+            await self._wait_for_run(thread_id, run.id)
+
+    async def _retrieve_last_run(self, thread_id: str) -> Optional[Run]:
+        list = await self._client.beta.threads.runs.list(thread_id, limit=1)
+        if len(list.data) > 0:
+            return list.data[0]
+        return None
+
+    def _is_run_completed(self, run: Run) -> bool:
+        if run.status == "completed":
+            return True
+        if run.status == "failed":
+            return True
+        if run.status == "cancelled":
+            return True
+        if run.status == "expired":
+            return True
+        return False
+
+    @staticmethod
+    async def create_assistant(
```

Comment:
can we move this public method up with the other publics? Usually easier to read if publics are above privates